// @hathor/ct-crypto-wasm/hathor_ct_crypto_wasm.js
/* @ts-self-types="./hathor_ct_crypto_wasm.d.ts" */

/**
 * Result of `rewindAmountShieldedOutput`. JS sees the fields as
 * `{ value: bigint, blindingFactor: Uint8Array }` via the auto-generated
 * getter / `toJSON` glue.
 */
export class RewoundAmountShieldedOutput {
    static __wrap(ptr) {
        const obj = Object.create(RewoundAmountShieldedOutput.prototype);
        obj.__wbg_ptr = ptr;
        RewoundAmountShieldedOutputFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        RewoundAmountShieldedOutputFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_rewoundamountshieldedoutput_free(ptr, 0);
    }
    /**
     * @returns {Uint8Array}
     */
    get blindingFactor() {
        const ret = wasm.rewoundamountshieldedoutput_blindingFactor(this.__wbg_ptr);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @returns {bigint}
     */
    get value() {
        const ret = wasm.rewoundamountshieldedoutput_value(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
}
if (Symbol.dispose) RewoundAmountShieldedOutput.prototype[Symbol.dispose] = RewoundAmountShieldedOutput.prototype.free;

/**
 * Result of `rewindFullShieldedOutput`. JS shape:
 * `{ value: bigint, blindingFactor: Uint8Array, tokenUid: Uint8Array, assetBlindingFactor: Uint8Array }`.
 */
export class RewoundFullShieldedOutput {
    static __wrap(ptr) {
        const obj = Object.create(RewoundFullShieldedOutput.prototype);
        obj.__wbg_ptr = ptr;
        RewoundFullShieldedOutputFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        RewoundFullShieldedOutputFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_rewoundfullshieldedoutput_free(ptr, 0);
    }
    /**
     * @returns {Uint8Array}
     */
    get assetBlindingFactor() {
        const ret = wasm.rewoundfullshieldedoutput_assetBlindingFactor(this.__wbg_ptr);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @returns {Uint8Array}
     */
    get blindingFactor() {
        const ret = wasm.rewoundfullshieldedoutput_blindingFactor(this.__wbg_ptr);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @returns {Uint8Array}
     */
    get tokenUid() {
        const ret = wasm.rewoundfullshieldedoutput_tokenUid(this.__wbg_ptr);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @returns {bigint}
     */
    get value() {
        const ret = wasm.rewoundfullshieldedoutput_value(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
}
if (Symbol.dispose) RewoundFullShieldedOutput.prototype[Symbol.dispose] = RewoundFullShieldedOutput.prototype.free;

/**
 * Build a blinded asset commitment (33-byte compressed point) from a raw
 * Tag and an asset blinding factor (`abf`). FullShielded value commitments
 * are computed against this generator instead of the unblinded asset tag.
 * @param {Uint8Array} tag_bytes
 * @param {Uint8Array} r_asset
 * @returns {Uint8Array}
 */
export function createAssetCommitment(tag_bytes, r_asset) {
    const ptr0 = passArray8ToWasm0(tag_bytes, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(r_asset, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.createAssetCommitment(ptr0, len0, ptr1, len1);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v3 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v3;
}

/**
 * Build a Pedersen commitment `C = amount * generator + blinding * G`.
 *
 * `amount` is passed as a JS `BigInt` (which arrives as `u64` after
 * wasm-bindgen's BigInt→u64 conversion); blinding is a 32-byte `vbf`;
 * generator is a 33-byte compressed point (asset tag for AmountShielded
 * or asset commitment for FullShielded).
 * @param {bigint} amount
 * @param {Uint8Array} blinding
 * @param {Uint8Array} generator
 * @returns {Uint8Array}
 */
export function createCommitment(amount, blinding, generator) {
    const ptr0 = passArray8ToWasm0(blinding, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(generator, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.createCommitment(amount, ptr0, len0, ptr1, len1);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v3 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v3;
}

/**
 * Build a trivial (zero-blinding) Pedersen commitment `C = amount * generator`.
 *
 * Mirrors the NAPI surface for completeness; not strictly needed by the
 * explorer's verify path but kept here so the two binding surfaces don't
 * diverge.
 * @param {bigint} amount
 * @param {Uint8Array} generator
 * @returns {Uint8Array}
 */
export function createTrivialCommitment(amount, generator) {
    const ptr0 = passArray8ToWasm0(generator, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.createTrivialCommitment(amount, ptr0, len0);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}

/**
 * Derive the deterministic asset-tag generator (33-byte compressed point) for a token UID.
 *
 * Used as the value-commitment generator for AmountShielded outputs (where
 * the token is public).
 * @param {Uint8Array} token_uid
 * @returns {Uint8Array}
 */
export function deriveAssetTag(token_uid) {
    const ptr0 = passArray8ToWasm0(token_uid, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.deriveAssetTag(ptr0, len0);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}

/**
 * Compute the ECDH shared secret used to derive the rewind nonce.
 *
 * Equivalent to libsecp256k1's `ecdh()` with the standard SHA-256 hash:
 * `SHA256(version_byte || x_coordinate)`. JS rarely needs this directly
 * — call `rewindAmountShieldedOutput` / `rewindFullShieldedOutput`
 * instead — but exposing it matches the NAPI surface and lets advanced
 * consumers compose their own flows.
 * @param {Uint8Array} private_key
 * @param {Uint8Array} peer_pubkey
 * @returns {Uint8Array}
 */
export function deriveEcdhSharedSecret(private_key, peer_pubkey) {
    const ptr0 = passArray8ToWasm0(private_key, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(peer_pubkey, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.deriveEcdhSharedSecret(ptr0, len0, ptr1, len1);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v3 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v3;
}

/**
 * Derive the raw 32-byte Tag for a token UID. Used when constructing
 * blinded asset commitments for FullShielded outputs.
 * @param {Uint8Array} token_uid
 * @returns {Uint8Array}
 */
export function deriveTag(token_uid) {
    const ptr0 = passArray8ToWasm0(token_uid, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.deriveTag(ptr0, len0);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}

/**
 * Return the HTR asset-tag generator (token_uid = [0; 32]).
 * @returns {Uint8Array}
 */
export function htrAssetTag() {
    const ret = wasm.htrAssetTag();
    var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v1;
}

/**
 * Rewind an AmountShielded output.
 *
 * Given the recipient's scan privkey + the on-chain output's
 * ephemeral pubkey + commitment + range proof + token UID, recovers
 * the cleartext `value` and `blindingFactor`. ECDH derives a shared
 * secret with the ephemeral pubkey; the secret seeds the range
 * proof's rewind nonce; the nonce unlocks the encrypted (value, vbf)
 * payload inside the proof.
 *
 * Returns `null` if the output isn't addressed to this scan key
 * (mismatched ECDH → wrong nonce → rewind fails). Throws on shape
 * errors (malformed inputs).
 * @param {Uint8Array} private_key
 * @param {Uint8Array} ephemeral_pubkey
 * @param {Uint8Array} commitment
 * @param {Uint8Array} range_proof
 * @param {Uint8Array} token_uid
 * @returns {RewoundAmountShieldedOutput}
 */
export function rewindAmountShieldedOutput(private_key, ephemeral_pubkey, commitment, range_proof, token_uid) {
    const ptr0 = passArray8ToWasm0(private_key, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(ephemeral_pubkey, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(commitment, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ptr3 = passArray8ToWasm0(range_proof, wasm.__wbindgen_malloc);
    const len3 = WASM_VECTOR_LEN;
    const ptr4 = passArray8ToWasm0(token_uid, wasm.__wbindgen_malloc);
    const len4 = WASM_VECTOR_LEN;
    const ret = wasm.rewindAmountShieldedOutput(ptr0, len0, ptr1, len1, ptr2, len2, ptr3, len3, ptr4, len4);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return RewoundAmountShieldedOutput.__wrap(ret[0]);
}

/**
 * Rewind a FullShielded output.
 *
 * Recovers value + blinding factor + token UID + asset blinding
 * factor. The token UID is encoded inside the proof's rewindable
 * message slot, so a single rewind returns all four. The
 * `assetCommitment` argument is the on-chain blinded asset
 * commitment for this output (33 bytes).
 * @param {Uint8Array} private_key
 * @param {Uint8Array} ephemeral_pubkey
 * @param {Uint8Array} commitment
 * @param {Uint8Array} range_proof
 * @param {Uint8Array} asset_commitment
 * @returns {RewoundFullShieldedOutput}
 */
export function rewindFullShieldedOutput(private_key, ephemeral_pubkey, commitment, range_proof, asset_commitment) {
    const ptr0 = passArray8ToWasm0(private_key, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(ephemeral_pubkey, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(commitment, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ptr3 = passArray8ToWasm0(range_proof, wasm.__wbindgen_malloc);
    const len3 = WASM_VECTOR_LEN;
    const ptr4 = passArray8ToWasm0(asset_commitment, wasm.__wbindgen_malloc);
    const len4 = WASM_VECTOR_LEN;
    const ret = wasm.rewindFullShieldedOutput(ptr0, len0, ptr1, len1, ptr2, len2, ptr3, len3, ptr4, len4);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return RewoundFullShieldedOutput.__wrap(ret[0]);
}
function __wbg_get_imports() {
    const import0 = {
        __proto__: null,
        __wbg___wbindgen_throw_9c31b086c2b26051: function(arg0, arg1) {
            throw new Error(getStringFromWasm0(arg0, arg1));
        },
        __wbindgen_cast_0000000000000001: function(arg0, arg1) {
            // Cast intrinsic for `Ref(String) -> Externref`.
            const ret = getStringFromWasm0(arg0, arg1);
            return ret;
        },
        __wbindgen_init_externref_table: function() {
            const table = wasm.__wbindgen_externrefs;
            const offset = table.grow(4);
            table.set(0, undefined);
            table.set(offset + 0, undefined);
            table.set(offset + 1, null);
            table.set(offset + 2, true);
            table.set(offset + 3, false);
        },
    };
    return {
        __proto__: null,
        "./hathor_ct_crypto_wasm_bg.js": import0,
    };
}

const RewoundAmountShieldedOutputFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_rewoundamountshieldedoutput_free(ptr, 1));
const RewoundFullShieldedOutputFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_rewoundfullshieldedoutput_free(ptr, 1));

function getArrayU8FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint8ArrayMemory0().subarray(ptr / 1, ptr / 1 + len);
}

function getStringFromWasm0(ptr, len) {
    return decodeText(ptr >>> 0, len);
}

let cachedUint8ArrayMemory0 = null;
function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}

function passArray8ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 1, 1) >>> 0;
    getUint8ArrayMemory0().set(arg, ptr / 1);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function takeFromExternrefTable0(idx) {
    const value = wasm.__wbindgen_externrefs.get(idx);
    wasm.__externref_table_dealloc(idx);
    return value;
}

let cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });
cachedTextDecoder.decode();
const MAX_SAFARI_DECODE_BYTES = 2146435072;
let numBytesDecoded = 0;
function decodeText(ptr, len) {
    numBytesDecoded += len;
    if (numBytesDecoded >= MAX_SAFARI_DECODE_BYTES) {
        cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });
        cachedTextDecoder.decode();
        numBytesDecoded = len;
    }
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}

let WASM_VECTOR_LEN = 0;

let wasmModule, wasmInstance, wasm;
function __wbg_finalize_init(instance, module) {
    wasmInstance = instance;
    wasm = instance.exports;
    wasmModule = module;
    cachedUint8ArrayMemory0 = null;
    wasm.__wbindgen_start();
    return wasm;
}

async function __wbg_load(module, imports) {
    if (typeof Response === 'function' && module instanceof Response) {
        if (typeof WebAssembly.instantiateStreaming === 'function') {
            try {
                return await WebAssembly.instantiateStreaming(module, imports);
            } catch (e) {
                const validResponse = module.ok && expectedResponseType(module.type);

                if (validResponse && module.headers.get('Content-Type') !== 'application/wasm') {
                    console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve Wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", e);

                } else { throw e; }
            }
        }

        const bytes = await module.arrayBuffer();
        return await WebAssembly.instantiate(bytes, imports);
    } else {
        const instance = await WebAssembly.instantiate(module, imports);

        if (instance instanceof WebAssembly.Instance) {
            return { instance, module };
        } else {
            return instance;
        }
    }

    function expectedResponseType(type) {
        switch (type) {
            case 'basic': case 'cors': case 'default': return true;
        }
        return false;
    }
}

function initSync(module) {
    if (wasm !== undefined) return wasm;


    if (module !== undefined) {
        if (Object.getPrototypeOf(module) === Object.prototype) {
            ({module} = module)
        } else {
            console.warn('using deprecated parameters for `initSync()`; pass a single object instead')
        }
    }

    const imports = __wbg_get_imports();
    if (!(module instanceof WebAssembly.Module)) {
        module = new WebAssembly.Module(module);
    }
    const instance = new WebAssembly.Instance(module, imports);
    return __wbg_finalize_init(instance, module);
}

async function __wbg_init(module_or_path) {
    if (wasm !== undefined) return wasm;


    if (module_or_path !== undefined) {
        if (Object.getPrototypeOf(module_or_path) === Object.prototype) {
            ({module_or_path} = module_or_path)
        } else {
            console.warn('using deprecated parameters for the initialization function; pass a single object instead')
        }
    }

    if (module_or_path === undefined) {
        module_or_path = new URL('hathor_ct_crypto_wasm_bg.wasm', import.meta.url);
    }
    const imports = __wbg_get_imports();

    if (typeof module_or_path === 'string' || (typeof Request === 'function' && module_or_path instanceof Request) || (typeof URL === 'function' && module_or_path instanceof URL)) {
        module_or_path = fetch(module_or_path);
    }

    const { instance, module } = await __wbg_load(await module_or_path, imports);

    return __wbg_finalize_init(instance, module);
}

export { initSync, __wbg_init as default };
